File: 01/puzzle_2.py
from collections import Counter


def solve(filepath):
  left = []
  right = []

  with open(filepath, "r") as file:
    for line in file:
      a, b = [int(x) for x in line.split()]

      left.append(a)
      right.append(b)

  counter = Counter(right)

  return sum(map(lambda x: x * counter[x], left))


result = solve("input.txt")

print(result)

File: 07/puzzle2.jl
using Pipe: @pipe

parse_input(filename) = @pipe (
  readlines(filename)
  .|> split(_, ':')
  .|> ((left, right),) -> (parse(Int, left), parse.(Int, split(strip(right))))
)

concat(a, b) = a * (10^length(digits(b))) + b

is_equation((test, numbers), total=0) =
  if total > test
    false
  elseif isempty(numbers)
    test == total
  else
    num, rest... = numbers

    is_eq(op) = is_equation((test, rest), op(total, num))
    any(is_eq, [concat, *, +])
  end

solve(filename) = @pipe(
  parse_input(filename)
  |> filter(is_equation)
  .|> first
  |> sum
)

solve("input.txt") |> println
